File: app/crawler/fetcher.py
"""HTTP fetching for the crawler: manual redirects, size limits, retries."""

from __future__ import annotations

import logging
import time
from collections.abc import Callable
from dataclasses import dataclass, field
from urllib.parse import urljoin

import httpx

from app.crawler.ssrf import SSRFError, is_ssrf_block, validate_url

logger = logging.getLogger(__name__)

RETRY_STATUSES = frozenset({429, 500, 502, 503, 504})
REDIRECT_STATUSES = frozenset({301, 302, 303, 307, 308})


class FetchError(Exception):
    def __init__(self, message: str, *, status: int | None = None) -> None:
        super().__init__(message)
        self.status = status


@dataclass
class FetchResult:
    requested_url: str
    final_url: str
    status: int
    headers: httpx.Headers
    body: bytes
    # (url, status, location) for every redirect hop that was followed or stopped at.
    redirects: list[tuple[str, int, str]] = field(default_factory=list)
    # Set when a redirect pointed somewhere we must not follow (other host, robots, SSRF).
    blocked_redirect: str | None = None

    @property
    def content_type(self) -> str:
        value = str(self.headers.get("content-type", ""))
        return value.split(";")[0].strip().lower()

    @property
    def is_html(self) -> bool:
        return self.content_type in ("text/html", "application/xhtml+xml")

    def text(self) -> str:
        charset = None
        for part in self.headers.get("content-type", "").split(";")[1:]:
            key, _, value = part.strip().partition("=")
            if key.lower() == "charset":
                charset = value.strip("\"' ")
        try:
            return self.body.decode(charset or "utf-8", errors="replace")
        except LookupError:
            return self.body.decode("utf-8", errors="replace")


@dataclass(frozen=True)
class FetcherConfig:
    user_agent: str
    max_retries: int = 2
    max_redirects: int = 5
    max_response_bytes: int = 5_000_000
    allowed_ports: tuple[int, ...] = (80, 443)
    max_retry_after_seconds: float = 10.0


class Fetcher:
    def __init__(
        self,
        client: httpx.Client,
        config: FetcherConfig,
        *,
        sleep: Callable[[float], None] = time.sleep,
    ) -> None:
        self._client = client
        self._config = config
        self._sleep = sleep

    def fetch(
        self,
        url: str,
        *,
        may_follow: Callable[[str], bool] = lambda _: True,
        accept: str = "text/html,application/xhtml+xml;q=0.9,*/*;q=0.5",
    ) -> FetchResult:
        """GET `url`, following redirects only to URLs accepted by `may_follow`."""
        redirects: list[tuple[str, int, str]] = []
        current = url
        for _ in range(self._config.max_redirects + 1):
            try:
                validate_url(current, allowed_ports=self._config.allowed_ports)
            except SSRFError as exc:
                raise FetchError(f"Blocked by SSRF protection: {exc}") from exc
            status, headers, body = self._get_with_retries(current, accept)
            location = headers.get("location")
            if status in REDIRECT_STATUSES and location:
                target = urljoin(current, location)
                redirects.append((current, status, target))
                blocked = False
                try:
                    validate_url(target, allowed_ports=self._config.allowed_ports)
                except SSRFError:
                    blocked = True
                if blocked or not may_follow(target):
                    return FetchResult(
                        url, current, status, headers, body, redirects, blocked_redirect=target
                    )
                current = target
                continue
            return FetchResult(url, current, status, headers, body, redirects)
        raise FetchError(f"Too many redirects (>{self._config.max_redirects})")

    def _get_with_retries(self, url: str, accept: str) -> tuple[int, httpx.Headers, bytes]:
        attempt = 0
        while True:
            retry_headers: httpx.Headers | None = None
            try:
                status, headers, body = self._get_once(url, accept)
            except httpx.TransportError as exc:
                if is_ssrf_block(exc):
                    raise FetchError(f"Blocked by SSRF protection: {exc}") from exc
                if attempt >= self._config.max_retries:
                    raise FetchError(f"{type(exc).__name__}: {exc}") from exc
                logger.info("Retrying %s after %s", url, type(exc).__name__)
            else:
                if status not in RETRY_STATUSES or attempt >= self._config.max_retries:
                    return status, headers, body
                retry_headers = headers
                logger.info("Retrying %s after HTTP %s", url, status)
            attempt += 1
            self._sleep(self._backoff(attempt, retry_headers))

    def _backoff(self, attempt: int, headers: httpx.Headers | None) -> float:
        retry_after = headers.get("retry-after") if headers is not None else None
        if retry_after and retry_after.strip().isdigit():
            return min(float(retry_after.strip()), self._config.max_retry_after_seconds)
        return min(2.0 ** (attempt - 1), self._config.max_retry_after_seconds)

    def _get_once(self, url: str, accept: str) -> tuple[int, httpx.Headers, bytes]:
        limit = self._config.max_response_bytes
        request_headers = {"User-Agent": self._config.user_agent, "Accept": accept}
        with self._client.stream(
            "GET", url, headers=request_headers, follow_redirects=False
        ) as response:
            declared = response.headers.get("content-length")
            if declared and declared.isdigit() and int(declared) > limit:
                raise FetchError(
                    f"Response too large ({declared} bytes > {limit})", status=response.status_code
                )
            chunks: list[bytes] = []
            size = 0
            if response.status_code not in REDIRECT_STATUSES:
                for chunk in response.iter_bytes():
                    size += len(chunk)
                    if size > limit:
                        raise FetchError(
                            f"Response too large (> {limit} bytes)", status=response.status_code
                        )
                    chunks.append(chunk)
            return response.status_code, response.headers, b"".join(chunks)

File: src/crawler/fetcher.js
/** HTTP fetching for the crawler: manual redirects, size limits, retries (app/crawler/fetcher.py). */

import { urljoin } from '../utils/pyurl.js';
import { pyStrip } from '../utils/pytext.js';
import { SSRFError, isSsrfBlock, ssrfMessage, validateUrl } from './ssrf.js';
import { logger } from '../utils/logger.js';

export const RETRY_STATUSES = new Set([429, 500, 502, 503, 504]);
export const REDIRECT_STATUSES = new Set([301, 302, 303, 307, 308]);
export const DEFAULT_ACCEPT = 'text/html,application/xhtml+xml;q=0.9,*/*;q=0.5';

export class FetchError extends Error {
  constructor(message, { status = null } = {}) {
    super(message);
    this.name = 'FetchError';
    this.status = status;
  }
}

/** A network-level failure (httpx.TransportError equivalent): retried, then reported. */
class TransportError extends Error {
  constructor(name, message, cause) {
    super(message);
    this.name = name;
    this.cause = cause;
  }
}

export class FetchResult {
  constructor(requestedUrl, finalUrl, status, headers, body, redirects = [], blockedRedirect = null) {
    this.requestedUrl = requestedUrl;
    this.finalUrl = finalUrl;
    this.status = status;
    this.headers = headers; // Headers (case-insensitive get)
    this.body = body; // Buffer
    // [url, status, location] for every redirect hop that was followed or stopped at.
    this.redirects = redirects;
    // Set when a redirect pointed somewhere we must not follow (other host, robots, SSRF).
    this.blockedRedirect = blockedRedirect;
  }

  get contentType() {
    return pyStrip(String(this.headers.get('content-type') ?? '').split(';')[0]).toLowerCase();
  }

  get isHtml() {
    return this.contentType === 'text/html' || this.contentType === 'application/xhtml+xml';
  }

  text() {
    let charset = null;
    for (const part of String(this.headers.get('content-type') ?? '').split(';').slice(1)) {
      const [key, , value] = partition(pyStrip(part), '=');
      if (key.toLowerCase() === 'charset') charset = value.replace(/^["' ]+|["' ]+$/g, '');
    }
    try {
      return new TextDecoder(charset || 'utf-8', { fatal: false }).decode(this.body);
    } catch {
      return new TextDecoder('utf-8', { fatal: false }).decode(this.body);
    }
  }
}

function partition(value, sep) {
  const i = value.indexOf(sep);
  return i < 0 ? [value, '', ''] : [value.slice(0, i), sep, value.slice(i + 1)];
}

export function fetcherConfig({
  userAgent,
  maxRetries = 2,
  maxRedirects = 5,
  maxResponseBytes = 5_000_000,
  allowedPorts = [80, 443],
  maxRetryAfterSeconds = 10.0,
}) {
  return { userAgent, maxRetries, maxRedirects, maxResponseBytes, allowedPorts, maxRetryAfterSeconds };
}

const sleepReal = (seconds) => new Promise((resolve) => setTimeout(resolve, seconds * 1000));

/**
 * @param fetchImpl `(url, init) => Promise<Response>`; production uses undici fetch with the
 *   SSRF-guarded dispatcher, tests inject a mock website.
 */
export class Fetcher {
  constructor(fetchImpl, config, { sleep = sleepReal } = {}) {
    this.fetchImpl = fetchImpl;
    this.config = config;
    this.sleep = sleep;
  }

  /** GET `url`, following redirects only to URLs accepted by `mayFollow`. */
  async fetch(url, { mayFollow = () => true, accept = DEFAULT_ACCEPT } = {}) {
    const redirects = [];
    let current = url;
    for (let i = 0; i < this.config.maxRedirects + 1; i += 1) {
      try {
        validateUrl(current, { allowedPorts: this.config.allowedPorts });
      } catch (err) {
        if (err instanceof SSRFError) throw new FetchError(`Blocked by SSRF protection: ${err.message}`);
        throw err;
      }
      const { status, headers, body } = await this.getWithRetries(current, accept);
      const location = headers.get('location');
      if (REDIRECT_STATUSES.has(status) && location) {
        const target = urljoin(current, location);
        redirects.push([current, status, target]);
        let isBlocked = false;
        try {
          validateUrl(target, { allowedPorts: this.config.allowedPorts });
        } catch (err) {
          if (!(err instanceof SSRFError)) throw err;
          isBlocked = true;
        }
        if (isBlocked || !mayFollow(target)) {
          return new FetchResult(url, current, status, headers, body, redirects, target);
        }
        current = target;
        continue;
      }
      return new FetchResult(url, current, status, headers, body, redirects);
    }
    throw new FetchError(`Too many redirects (>${this.config.maxRedirects})`);
  }

  async getWithRetries(url, accept) {
    let attempt = 0;
    for (;;) {
      let retryHeaders = null;
      try {
        const result = await this.getOnce(url, accept);
        if (!RETRY_STATUSES.has(result.status) || attempt >= this.config.maxRetries) return result;
        retryHeaders = result.headers;
        logger.info(`Retrying ${url} after HTTP ${result.status}`);
      } catch (err) {
        if (!(err instanceof TransportError)) throw err;
        if (isSsrfBlock(err)) throw new FetchError(`Blocked by SSRF protection: ${ssrfMessage(err)}`);
        if (attempt >= this.config.maxRetries) throw new FetchError(`${err.name}: ${err.message}`);
        logger.info(`Retrying ${url} after ${err.name}`);
      }
      attempt += 1;
      await this.sleep(this.backoff(attempt, retryHeaders));
    }
  }

  backoff(attempt, headers) {
    const retryAfter = headers ? headers.get('retry-after') : null;
    if (retryAfter && /^\d+$/.test(pyStrip(retryAfter))) {
      return Math.min(Number(pyStrip(retryAfter)), this.config.maxRetryAfterSeconds);
    }
    return Math.min(2.0 ** (attempt - 1), this.config.maxRetryAfterSeconds);
  }

  async getOnce(url, accept) {
    const limit = this.config.maxResponseBytes;
    let response;
    try {
      response = await this.fetchImpl(url, {
        method: 'GET',
        redirect: 'manual',
        headers: { 'User-Agent': this.config.userAgent, Accept: accept },
      });
    } catch (err) {
      throw toTransportError(err);
    }
    const status = response.status;
    const declared = response.headers.get('content-length');
    if (declared && /^\d+$/.test(declared) && Number(declared) > limit) {
      await cancelBody(response);
      throw new FetchError(`Response too large (${declared} bytes > ${limit})`, { status });
    }
    if (REDIRECT_STATUSES.has(status)) {
      await cancelBody(response);
      return { status, headers: response.headers, body: Buffer.alloc(0) };
    }
    const chunks = [];
    let size = 0;
    if (response.body) {
      const reader = response.body.getReader();
      try {
        for (;;) {
          const { done, value } = await reader.read();
          if (done) break;
          size += value.byteLength;
          if (size > limit) {
            await reader.cancel().catch(() => {});
            throw new FetchError(`Response too large (> ${limit} bytes)`, { status });
          }
          chunks.push(Buffer.from(value));
        }
      } catch (err) {
        if (err instanceof FetchError) throw err;
        throw toTransportError(err);
      }
    }
    return { status, headers: response.headers, body: Buffer.concat(chunks) };
  }
}

async function cancelBody(response) {
  try {
    await response.body?.cancel();
  } catch {
    /* ignore */
  }
}

/** Map fetch/undici failures to named transport errors (like httpx exception class names). */
function toTransportError(err) {
  if (err instanceof TransportError || err instanceof FetchError) return err;
  const root = err?.cause ?? err;
  const code = root?.code ?? '';
  let name = 'ConnectError';
  if (err?.name === 'TimeoutError' || /TIMEOUT/i.test(code)) name = 'TimeoutException';
  else if (/ECONNRESET|UND_ERR_SOCKET|EPIPE/.test(code)) name = 'ReadError';
  const message = root?.message || err?.message || String(err);
  return new TransportError(name, message, err);
}
